pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountThetaSketchAggregationFunction.java

```diff
@@ -184,346 +192,1063 @@ public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int ma
   @Override
   public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
       Map<ExpressionContext, BlockValSet> blockValSetMap) {
-    Map<Predicate, Union> unionMap = getUnionMap(aggregationResultHolder);
-
-    Sketch[] sketches = deserializeSketches(blockValSetMap.get(_thetaSketchColumn).getBytesValuesSV(), length);
-    for (PredicateInfo predicateInfo : _predicateInfoMap.values()) {
-      Predicate predicate = predicateInfo.getPredicate();
-      BlockValSet blockValSet = blockValSetMap.get(predicate.getLhs());
-      FieldSpec.DataType valueType = blockValSet.getValueType();
-      PredicateEvaluator predicateEvaluator = predicateInfo.getPredicateEvaluator(valueType);
-
-      Union union = unionMap.get(predicate);
-      switch (valueType) {
-        case INT:
-          int[] intValues = blockValSet.getIntValuesSV();
-          for (int i = 0; i < length; i++) {
-            if (predicateEvaluator.applySV(intValues[i])) {
-              union.update(sketches[i]);
+    int numExpressions = _inputExpressions.size();
+    boolean[] singleValues = new boolean[numExpressions];
+    DataType[] valueTypes = new DataType[numExpressions];
+    Object[] valueArrays = new Object[numExpressions];
+    extractValues(blockValSetMap, singleValues, valueTypes, valueArrays);
+    int numFilters = _filterEvaluators.size();
+
+    // Main expression is always index 0
+    if (valueTypes[0] != DataType.BYTES) {
+      List<UpdateSketch> updateSketches = getUpdateSketches(aggregationResultHolder);
+      if (singleValues[0]) {
```

Comment:
Are these if/else blocks re-factorable?

Reply:
I tried and don't see a better way to organize the code. We need to have slightly different logic for each primitive type